File: kotlin/run-length-encoding/src/main/kotlin/RunLengthEncoding.kt
object RunLengthEncoding {

    fun encode(input: String): String =
        input.replace("(.)\\1+".toRegex()) { "${it.value.length}${it.groupValues[1]}" }

    fun decode(input: String): String =
        input.replace("(\\d+)(.)".toRegex()) { it.groupValues[2].repeat(it.groupValues[1].toInt()) }
}

File: clojure/run-length-encoding/src/run_length_encoding.clj
(ns run-length-encoding
  (:require [clojure.string :as str]))

(defn run-length-encode
  "encodes a string with run-length-encoding"
  [plain-text]
  (str/replace 
    plain-text 
    #"(.)\1+" 
    #(str (str (count (% 0))) (str (% 1)))))

(defn run-length-decode
  "decodes a run-length-encoded string"
  [cipher-text]
  (str/replace 
    cipher-text
    #"(\d+)(.)" 
    #(apply str (repeat (Integer/parseInt (% 1)) (str (% 2))))))
